### PR TITLE
refactor: use debug .. in for filter expression

### PR DESCRIPTION
**Pull request by tonyfettes**
# Design

```
debug eval($e + $e) in
^^^^^               ^^ keyword in Form.re
```

The expression above is parsed into

```reason
Filter(Ap(Var("eval"), BinOp(Plus, Unquote(Var("e")), Unquote(Var("e"))))
```

### CODE REVIEW

**Comment by tonyfettes** [comment]
Note, editing the filter will not update the stepper/filter behavior immediately, and has to be update via an explicit re-activation of the stepper, i.e. turning off the stepper history then turning it on again.

**Comment by cyrus-** [comment]
@tonyfettes you'll need to update the tests (and consider adding some additional tests) -- currently 4 tests are failing

**Comment by cyrus-** [comment]
Can we also make a couple of documentation slides with examples, collectively covering each of the 4 commands? Ideally interesting examples, not just arithmetic. You can use this one for eval + stop:

image

**Comment by cyrus-** [comment]
Also update the explainthis for debug with examples of the other two commands.

**Comment by cyrus-** [comment]
This documentation could also use an update: 
image.

**Comment by tonyfettes** [comment]
## Guide to this PR

TL;DR: This PR replaces the old filter syntax `eval/hide/debug/pause <exp> in ..` with `debug eval/hide/stop/step(<exp>) in ..`.

This PR consists of following major changes:

1. Change the AST representation of $e and $v from a general `Constructor("$e")` and `Constructor("$v")` to special AST node:
   - `$e`: `FilterSelector(Exp)`
   - `$v`: `FilterSelector(Val)`
   
   So that the typing of these two pattern is not coupled with constructor typings, and we don't have to inject a artificially manufactured sum type in order for `$e` and `$v` to be typed.
2. Refactor the parsing rules for debug filter language constructs from

   ```
   <debug-filter> ::= eval <exp> in
                   |  hide <exp> in
                   |  debug <exp> in
                   |  pause <exp> in
   ```
   
   to
   
   ```
   <debug-filter> ::= debug <exp> in
                  where <exp> is of the form <var>(<exp>)
                  where <var> is "eval"/"hide"/"stop"/"step"
   ```
3. Introduces the concept of "Unresolved Filter" (`Filter(Unresolved(..), ..)`) and "Resolved Filter" (`Filter(Filter(..), ..)`). The parser emits the unresolved form of filter, and it will be resolved by the type-checker (Statics.re), so that we can extract ids (location information) out of the expression. We used to directly resolve the filter in the parser, which makes it impossible to send corresponding ids to other parts like ExplainThis.
4. Bug fixes to the dynamics of the filter, focusing mostly on what should be skipped and what should not. Specifically, this PR:
   - Use a equality check ignoring the ascription only.
   - Check for CompletedFilter before testing if the step should be skipped by user settings.
5. Major documentation & example works including:
   - Improvement on ExplainThis over the debug filter statements.
   - A slide deck showing some usage of debug filter.

**Comment by tonyfettes** [comment]
---
schemaVersion: 1
prNumber: 1388
prOwner: hazelgrove
prRepo: hazel
baseSha: fca48094c3a00e5ed97fee603eb820aac148ca9d
headSha: e7930d6993876e35c49fbea45e775429e56b16be
---

# refactor: use debug .. in for filter expression

This PR replaces the four separate filter keyword forms (`hide`, `eval`, `pause`, `debug`) with a single unified `debug action(pattern) in body` surface syntax. It also introduces two first-class expression variants — `FilterAction` and `FilterSelector` — so filter actions and the `$e`/`$v` selectors are proper typed expressions rather than magic constructor strings. The change touches the grammar, parser/lexer, statics, dynamics, pretty-printer, ExplainThis docs, and the test suite.

## New expression variants: `FilterAction` and `FilterSelector`

Previously, filter actions like `eval` and `pause` were represented as `Constructor("$e", …)` strings in the expression tree — a fragile convention. This refactor promotes them to proper first-class variants `FilterAction` and `FilterSelector` in `Grammar.re`'s `exp_term` type and `FilterSelector.re`'s new module. The `filter` record also gains an `ids` field to carry annotation data through elaboration. The `Unresolved` constructor is added to `stepper_filter_kind_t` to hold the not-yet-resolved parsed expression before statics lifts it into a resolved `Filter`.

<details open>
<summary><code>src/language/term/FilterSelector.re</code> · [@deriving (show({with_path: false}), sexp, yojson, eq)]</summary>



```diff
@@ -0,0 +1,4 @@
+[@deriving (show({with_path: false}), sexp, yojson, eq)]
+type t =
+  | Exp
+  | Val;
```

</details>

<details open>
<summary><code>src/language/term/Grammar.re</code> · | FilterAction(FilterAction.t)</summary>



```diff
@@ -48,6 +48,8 @@ and exp_term('a) =
   | Dot(exp_t('a), exp_t('a))
   | LivelitName(string)
   | Var(Var.t)
+  | FilterAction(FilterAction.t)
+  | FilterSelector(FilterSelector.t)
   | Let(pat_t('a), exp_t('a), exp_t('a))
   | Theorem(pat_t('a), exp_t('a), exp_t('a))
   | ProofObject(exp_t('a))
```

</details>

## Statics: resolve & typing debug filter

We added a new Unresolved variant to Filter so that we can postpone the resolution (pattern-matching, to be more precise) on the expression inside `debug <expr> in ...` and get the action and the filter pattern out, and populating the info map properly, so that facilities like ExplainThis can correctly display information about these language constructs.

<details open>
<summary><code>src/language/term/Grammar.re</code> · | Unresolved(exp_t('a))</summary>



```diff
@@ -152,6 +154,7 @@ and mpat_term('a) =
   | Asc(mpat_t('a), typ_t('a))
 and mpat_t('a) = Annotated.t(mpat_term('a), 'a)
 and stepper_filter_kind_t('a) =
+  | Unresolved(exp_t('a))
   | Filter(filter('a))
   | Residue(int, FilterAction.t)
 and type_hole('a) =
```

</details>

<details open>
<summary><code>src/language/term/Grammar.re</code> · ids: 'a,</summary>



```diff
@@ -165,6 +168,7 @@ and type_provenance('a) =
 and filter('a) = {
   pat: exp_t('a),
   act: FilterAction.t,
+  ids: 'a,
 };
```

</details>

<details open>
<summary><code>src/language/statics/Statics.re</code> · | Filter(Filter({pat: cond, act}), body) =&gt;</summary>



```diff
@@ -1451,7 +1467,44 @@ and uexp_to_info_map =
           SubexpProbeTargets.union_all([e.probe_targets, hint.probe_targets]),
         m,
       );
-    | Filter(Filter({pat: cond, act}), body) =>
+    | Filter(Unresolved(fexp), body) =>
+      /* Recognize the parsed Ap form `act(pat)` as a resolved filter; lift
+         it into Filter(Filter({...})). Anything else stays Unresolved
+         (typed as syn for display) and we still descend into body. */
+      switch (fexp.term) {
+      | Ap(Forward, {term: FilterAction(act), annotation: act_ann}, pat) =>
+        let (_, _, m) = go(~ana=syn, fexp, m, ~is_in_filter=true);
+        let (cond, cond_elab, m) = go(~ana=syn, pat, m, ~is_in_filter=true);
+        let (body, body_elab, m) = go(~ana, body, m);
+        add(
+          ~elab_term=
+            Filter(
+              Filter({
+                act,
+                pat: cond_elab,
+                ids: act_ann,
+              }),
+              body_elab,
+            )
+            |> rewrap,
+          ~elab_syn_ty=body.elab_syn_ty,
+          ~marks=[],
+          ~co_ctx=CoCtx.union([cond.co_ctx, body.co_ctx]),
+          m,
+        );
+      | _ =>
+        let (fexp_info, fexp_elab, m) =
+          go(~ana=syn, fexp, m, ~is_in_filter=true);
+        let (body, body_elab, m) = go(~ana, body, m);
+        add(
+          ~elab_term=Filter(Unresolved(fexp_elab), body_elab) |> rewrap,
+          ~elab_syn_ty=body.elab_syn_ty,
+          ~marks=[],
+          ~co_ctx=CoCtx.union([fexp_info.co_ctx, body.co_ctx]),
+          m,
+        );
+      }
+    | Filter(Filter({pat: cond, act, ids: filter_ids}), body) =>
       let (cond, cond_elab, m) = go(~ana=syn, cond, m, ~is_in_filter=true);
       let (body, body_elab, m) = go(~ana, body, m);
       add(
```

</details>

## Parser & lexer: unified `debug .. in` syntax

The four separate keyword tokens (`PAUSE`, `HIDE`, `EVAL`, `DEBUG`) and their grammar rules collapse into a single `DEBUG … IN` form. The action and pattern are now ordinary expressions — `action(pattern)` — so the parser simply parses them as an expression argument. Two new lexer tokens `$e` and `$v` are added for the `FilterSelector` values, and the `AST.re` representation of `Filter` drops the explicit `filter_action` constructor argument.

<details open>
<summary><code>src/menhirParser/Lexer.mll</code> · (* Filter Selectors *)</summary>



```diff
@@ -98,6 +98,9 @@ rule token =
     | "!=." { NOT_EQUAL_FLOAT }
     (* String Ops *)
     | "++" { STRING_CONCAT }
+    (* Filter Selectors *)
+    | "$e" { FILTER_SELECTOR_EXP }
+    | "$v" { FILTER_SELECTOR_VAL }
     (* Bool ops *)
     | "&&" { L_AND }
     | "||" { L_OR }
```

</details>

<details open>
<summary><code>src/menhirParser/Lexer.mll</code> · | "pause" {PAUSE}</summary>



```diff
@@ -116,10 +119,7 @@ rule token =
     (* DHExp Annotations *)
     | "()" { UNIT }
     (* Filters *)
-    | "pause" {PAUSE}
     | "debug" {DEBUG}
-    | "hide" {HIDE}
-    | "eval" {EVAL}
     (* Other *)
     | ";" {SEMI_COLON}
     | "test" {TEST}
```

</details>

<details open>
<summary><code>src/menhirParser/Parser.mly</code> · %token PAUSE</summary>



```diff
@@ -23,10 +23,7 @@ open AST
 %token TYP_AP_SYMBOL
 %token CONS
 %token TEST
-%token PAUSE
 %token DEBUG
-%token HIDE
-%token EVAL
 %token <string> IDENT
 %token <string> CONSTRUCTOR_IDENT
 %token <string> STRING
```

</details>

<details open>
<summary><code>src/menhirParser/Parser.mly</code> · (* filter selector *)</summary>



```diff
@@ -110,6 +107,9 @@ open AST
 
 %token SEMI_COLON
 
+(* filter selector *)
+%token FILTER_SELECTOR_EXP
+%token FILTER_SELECTOR_VAL
 
 
 (* Precedences *)
```

</details>

<details open>
<summary><code>src/menhirParser/Parser.mly</code> · filterAction:</summary>



```diff
@@ -314,12 +314,6 @@ funExp:
 %inline ifExp:
     | IF; e1 = exp; THEN; e2 = exp; ELSE; e3 = exp { If (e1, e2, e3) } %prec IF_EXP
 
-filterAction:
-    | PAUSE { Pause }
-    | DEBUG { Debug }
-    | HIDE { Hide }
-    | EVAL { Eval }
-
 tpat:
     | TP_TPAT; s = STRING {InvalidTPat(s)}
     | p = PROJECTOR_INVOKE {InvalidTPat(p)}
```

</details>

<details open>
<summary><code>src/menhirParser/Parser.mly</code> · | a = filterAction; cond = exp; IN; body = exp { Filter(a,…</summary>



```diff
@@ -369,7 +363,9 @@ exp:
     | FIX;  p = funPat; DASH_ARROW; e = exp { FixF(p, e) }
     | TYP_FUN; t = tpat; DASH_ARROW; e = exp {TypFun(t, e)}
     | QUESTION { EmptyHole }
-    | a = filterAction; cond = exp; IN; body = exp { Filter(a, cond, body)} %prec LET_EXP
+    | DEBUG; cond = exp; IN; body = exp { Filter(cond, body) } %prec LET_EXP
+    | FILTER_SELECTOR_EXP { FilterSelector(Exp) }
+    | FILTER_SELECTOR_VAL { FilterSelector(Val) }
     | TEST; e = exp; END { Test(e) }
     | e1 = exp; AT_SYMBOL; e2 = exp { ListConcat(e1, e2) }
     | e1 = exp; CONS; e2 = exp { Cons(e1, e2) }
```

</details>

<details open>
<summary><code>src/menhirParser/AST.re</code> · | FilterAction(Language.FilterAction.t)</summary>



```diff
@@ -138,6 +162,8 @@ and deferral_pos =
 and exp =
   | Atom(Language.Atom.t)
   | Var(string)
+  | FilterAction(Language.FilterAction.t)
+  | FilterSelector(Language.FilterSelector.t)
   | Constructor(string, option(option(typ)))
   | ListExp(list(exp))
   | TupleExp(list(exp))
```

</details>

<details open>
<summary><code>src/menhirParser/AST.re</code> · | Filter(filter_action, exp, exp)</summary>



```diff
@@ -157,7 +183,7 @@ and exp =
   | FixF(pat, exp)
   | Asc(exp, typ)
   | EmptyHole
-  | Filter(filter_action, exp, exp)
+  | Filter(exp, exp)
   | BuiltinFun(string)
   | Undefined
   | Seq(exp, exp)
```

</details>

## Dynamics: CompleteFilter fix and user settings override.

We here first check if the eval object is already CompleteFilter, and just skip over it; otherwise we will get more and more residue and can never return. Also here we adjust the guard expression, so  that user settings will be properly respected in the stepper.

<details open>
<summary><code>src/language/dynamics/stepper/EvaluatorStep.re</code> · (~settings, x: EvalObj.t): (FilterAction.action, EvalObj.t)…</summary>



```diff
@@ -213,29 +216,34 @@ let rec matches =
 };
 
 let should_hide_eval_obj =
-    (~settings, x: EvalObj.t): (FilterAction.action, EvalObj.t) =>
-  if (should_hide_step_kind(~settings, x.knd)) {
-    (Eval, x);
-  } else {
-    let (act, _, ctx) =
+    (~settings, x: EvalObj.t): (FilterAction.action, EvalObj.t) => {
+  switch (x.knd) {
+  | CompleteFilter =>
+    /* CompleteFilter removes internal filter bookkeeping. Running filter
+       matching on this step can re-introduce a Residue wrapper around the
+       expression it just unwrapped, causing an infinite sequence of hidden
+       CompleteFilter steps. */
+    (Eval, x)
+  | _ =>
+    let (act, idx, ctx) =
       matches(Environment.empty, [], x.ctx, x.d_loc, (Step, One), 0);
+    let x = {
+      ...x,
+      ctx,
+    };
     switch (act) {
-    | (Eval, _) => (
+    | (Eval, _) => (Eval, x)
+    // Skip over Ascription
+    | (Step, _) when step_kind_is_unrenderable(~settings, x.knd) => (
         Eval,
-        {
-          ...x,
-          ctx,
-        },
-      )
-    | (Step, _) => (
-        Step,
-        {
-          ...x,
-          ctx,
-        },
+        x,
       )
+    | (Step, _) when idx > 0 => (Step, x)
+    | (Step, _) when should_hide_step_kind(~settings, x.knd) => (Eval, x)
+    | (Step, _) => (Step, x)
     };
   };
+};
 
 module Decompose = {
   module Result = {
```

</details>
